### ISSUE TITLE
HTTPS redirect is hard coded to 301 MOVED PERMANENTLY

### ISSUE THREAD

**Issue by andrewmackett** [terraform-google-modules/terraform-google-lb-http#236]
### TL;DR

The HTTPS redirect is hard coded to 301 MOVED PERMANENTLY which allows the request method to be changed from POST to GET.

### Expected behavior

The status code for the HTTPS redirect should be configurable by a variable to allow the use of any status code supported by the `compute_url_map` resource:

https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_url_map#redirect_response_code

### Observed behavior

All HTTPS redirects use the 301 MOVED PERMANENTLY status code.

### Terraform Configuration

```hcl
# https://github.com/terraform-google-modules/terraform-google-lb-http/blob/master/main.tf#L146-L155

resource "google_compute_url_map" "https_redirect" {
  project = var.project
  count   = var.https_redirect ? 1 : 0
  name    = "${var.name}-https-redirect"
  default_url_redirect {
    https_redirect         = true
    redirect_response_code = "MOVED_PERMANENTLY_DEFAULT"
    strip_query            = false
  }
}
```

### Terraform Version

```sh
Terraform v1.0.10
```

### Additional information

I've opened PR https://github.com/terraform-google-modules/terraform-google-lb-http/pull/235 to add a variable to allow the status code used in the HTTPS redirect to be configured.

**Comment by andrewmackett** [comment]
@morgante the docs you linked to says (emphasis mine):
> Setting the Enable HTTP to HTTPS Redirect checkbox automatically creates an additional partial HTTP load balancer consisting of a URL map, a forwarding rule, and a target HTTP proxy. This partial HTTP load balancer uses the same IP address as your HTTPS load balancer and redirects HTTP requests to your load balancer's HTTPS frontend. __It uses 301 Moved Permanently as the default redirect response code.__

The change I proposed in PR https://github.com/terraform-google-modules/terraform-google-lb-http/pull/235 also continues to use 301 as the default redirect response code.

Further down the [same document](https://cloud.google.com/load-balancing/docs/https/setting-up-http-https-redirect#partial-http-lb) is says (again, emphasis mine):
> This example uses the 301 response code. __You can instead use a different [response code](https://cloud.google.com/load-balancing/docs/https/traffic-management#response-codes).__

It links to the [supported response codes](https://cloud.google.com/load-balancing/docs/https/traffic-management#response-codes):

| Response Code | Number | Notes |
|---------------|--------|-------|  
| MOVED_PERMANENTLY_DEFAULT | 301 |  |
| FOUND | 302 |  |
| PERMANENT_REDIRECT | 308 | In this case, the request method is retained.|
| TEMPORARY_REDIRECT | 307 | In this case, the request method is retained.|
| SEE_OTHER | 303 | |

I think exposing this variable is important for situations where you want to maintain the request method after the redirect. I agree that if the backend is serving a website a 301 would be most suitable. However if the backend is serving an API I think it would be better to be able to use a 307 or 308 to ensure the client maintains the request method after the redirect.

**Comment by morgante** [comment]
Yes, I understand that other options are possible. But they are not desirable. In my view, the http->https redirect is mainly a convenience for users who manually type in a URL (or click a link). Any application/client which is sending POST requests needs to use the https URL, as doing otherwise risks leaking un-encrypted data (the initial request is un-encrypted). Really, any application which is using the HTTP url needs to be updated—you should not rely on the redirects.

**Comment by morgante** [comment]
>  I thought that the flexibility to use 307 or 308 would have been useful to cover accidental use of http - but perhaps it is better for the redirected request to fail because of the different request method.

Yeah, ultimately since a POST request is machine-initiated (not human-initiated) I think it's better for it to fail. You don't want applications to continue relying on the redirect and thereby leaking data.

> I'm happy to close this issue if you don't wish to support other redirect status codes. Thanks.

Thanks for understanding, and really do appreciate the contribution here.
